### PR TITLE
`git stash ...` with Clink v1.1.12 popup list

### PR DESCRIPTION
**Pull request by chrisant996**
This is meant to be a PR for 0206ae1 and 262106d.  It might cite 3 other commits as well, but they should amount to no diffs as they were just the same change made in two branches and then merging with itself.  I had been working out of origin/master originally, and I think that led to some muddled tracking.  I've changed to using topic branches.

Anyway, these two changes use a new Clink API (after testing what version is being used) to do match display filtering so that stash descriptions can show up in the `clink-popup-complete` command (and also so stash names vs descriptions are colored differently in the console).

Both changes are written to be compatible with all versions of Clink.

Around line 262 | 269 there might be a conflict with one of the other PRs I submitted recently.  The merge should be simple; keep the block with the if/elseif/else/end block.

(UPDATE: I resolved the merged conflict a minute or two after originally posting this comment.)

### CODE REVIEW

**Comment by vladimir-kotikov** [comment]
@chrisant996 thanks for your contribution. Could you please fix CI before I merge this in

**Comment by chrisant996** [comment]
I'll be happy to fix anything that needs fixing.

But what needs to be fixed?

If it's the 3 appveyor failures, those are false positives. The `rl` and `settings` variables are both defined by Clink's C++ code, and the accesses are also behind Clink version checks. I've manually verified that the change runs properly on both new Clink and old Clink (0.4.9).

I'm unfamiliar with how this appveyor build works. How can I teach it about these variables so it doesn't flag them as failures?

**Comment by chrisant996** [comment]
> I'm unfamiliar with how this appveyor build works. How can I teach it about these variables so it doesn't flag them as failures?

For example, why aren't the usages of `clink` flagged as using undefined variables?  The `clink` and `rl` and `settings` variables are all defined by the same C++ code in Clink.

I can't find anywhere in the .appveyor.yml file or anywhere else in the repo that resolves the false positives, so I'm a little confused how to satisfy appveyor.

**Comment by vladimir-kotikov** [comment]
I think what you would need to is to add those `clink`-injected variables to this list so that they are properly detected as 'global'

**Comment by chrisant996** [comment]
> I think what you would need to is to add those `clink`-injected variables to this list so that they are properly detected as 'global'

Sorry, I'm confused.  To which list do I need to add them?  I can't find such a list.  I apologize in advance if I'm just being blind.

**Comment by vladimir-kotikov** [comment]
Shit, I forgot to paste a link to the file :) Sorry for confusion. This is the line you're seeking for: https://github.com/vladimir-kotikov/clink-completions/blob/master/.luacheckrc#L6

**Comment by chrisant996** [comment]
Ah! All is clear now, thanks. Will update it.

**Comment by chrisant996** [comment]
Done; CI passes now.

**Comment by vladimir-kotikov** [comment]
@chrisant996 thanks!

File: git.lua
Comment (vladimir-kotikov) [inline git.lua:163]: There's already a coupple of functions in [`color.lua`](https://github.com/vladimir-kotikov/clink-completions/blob/master/modules/color.lua) to enable color output - could you please reuse those functions (possibly updating them to add extra functionality if needed) instead of doing manual string concatenations?

File: git.lua
Comment (vladimir-kotikov) [inline git.lua:273]: This is the second time this magic number appears which makes me thinking that it's worth at least moving it to a dedicated constant (along with `10010012` two lines above) but in general I'd love to see a separate module for clink version/feature detection (I think it's likely we'll see more of these version constants sooner or later)

File: git.lua
Comment (vladimir-kotikov) [inline git.lua:272]: Did clink interface to add completions change in this version? If so why does this apply only to `git stash` completions? How does this affect other commands completions?

File: git.lua
Comment (vladimir-kotikov) [inline git.lua:274]: You're resetting color to default here (`\x1b[m` sequence is about that IIRC) but I can't see where the color been chnaged from default

**Review by vladimir-kotikov — CHANGES_REQUESTED** [review]

**Comment by vladimir-kotikov** [comment]
@chrisant996 apologize for being nitpicky and getting back and forth on this but I'd like to figure out a couple thing before merging this PR.

File: git.lua
Comment (chrisant996) [reply git.lua:274]: Clink itself can set the color (per the `color.filtered` setting).  The reset here forces normal text for the description.  The effect is the stash name in a user-defined color (default is "bold") and the stash description in another color, the default color.  So the default effect is bright text for the name and dim text for the description.

I can add a comment to make it clear.

File: git.lua
Comment (chrisant996) [reply git.lua:272]: Yes, big changes.  Backward compatibility exists, so the existing stuff all continues to work.  The changes here make use of new functionality when available.

File: git.lua
Comment (chrisant996) [reply git.lua:273]: The format is documented in the Clink documentation.  E.g. `10020034` is v1.2.34.
The values are easily looked up in the Changes section of the documentation.

It's intended as a lightweight capability interrogation mechanism.  I'd love to have a more sophisticated system, but designing it takes careful planning.  I don't think it makes sense to block the release of Clink over that.  This lightweight mechanism covers all the bases from a functional perspective, but isn't pretty.

To respect the clink-completions code base, I can build a module for clink-completions to make version queries use friendly function names.

File: git.lua
Comment (chrisant996) [reply git.lua:163]: Ok.

Comment (chrisant996) [reply git.lua]: Done.

File: git.lua
Comment (chrisant996) [reply git.lua]: Added a version module.

File: git.lua
Comment (chrisant996) [reply git.lua]: I took the simpler approach of only removing the support for v1.1.9 through v1.1.11 since they're testing builds and v1.1.12 has further enhancements.  The version checks are encapsulated in the version module.

**Comment by chrisant996** [comment]
> @chrisant996 apologize for being nitpicky and getting back and forth on this but I'd like to figure out a couple thing before merging this PR.

It's ok. I like a rigorous approach, when it adds value.

The CI stuff is frustrating, though:

- The modules\*.lua scripts are executed _outside of Clink_. So if they try to use anything from Clink, then they fail.
- I like the idea of encapsulating new Clink color functionality into the color module, and encapsulating new Clink version checks into a version module.
- But since the modules aren't allowed to run any Clink code, that makes it pretty hard to satisfy code coverage analysis!

How would you suggest to proceed?
- Should I try to trick the code coverage analysis by using the short circuiting `and` operator to put a version check on every line and rely on code coverage not noticing that most of the line is skipped?
- Should I move the new `get_clink_color` function from color.lua into version.lua, and exclude version.lua from code coverage?
- Is there a way to bracket some lines with some kind of "exclude from code coverage" annotation?
- Should I build a mock of Clink's functionality to trick the code coverage analysis? If so, where can I put it so that it's only used during code coverage testing and isn't included during normal installation?

**Comment by chrisant996** [comment]
I essentially tricked the code coverage, because I'm not willing to invest in building a mock infrastructure to mimic Clink's API.

File: git.lua
Comment (vladimir-kotikov) [reply git.lua]: I understand your concerns but I prefer to have readable and meaningful constant names rather than 'cryptic' numbers so that whoever is reading the code they don't need to look up in the changelog and can figure out what this check is about by just reading the constant name

**Comment by vladimir-kotikov** [comment]
I'd rather just ignore the coverage and just add tests/mock things only if you feel that it brings any real value and doesn't worth too much time

File: git.lua
Comment (vladimir-kotikov) [inline git.lua:11]: Could you rename the variable (and perhaps module as well) to `clink_version` - IMHO `clink_version.supports_color_settings()` reads slightly more natural

File: modules/version.lua
Comment (vladimir-kotikov) [inline modules/version.lua:9]: Does this (and the other exports as well) needs to be a function? Seems like it can be a plain variable which will be evaluated at script execution time when clink-injected variables are already available

File: modules/color.lua
Comment (vladimir-kotikov) [inline modules/color.lua:29]: I wonder what would `settings.get("color.git.star")` return? Is it single color number from [this table](https://en.wikipedia.org/wiki/ANSI_escape_code#SGR_parameters) or complete EGR-sequence or something  else? I assume it should be either 1) or 2) but just wanted to make sure.

Also in case it's a single EGR (color) code,  should there be a separate function to convert color code to full escape sequence? Can't you simply do something like `colors.color_text("foo", settings.get("color.git.star"))` instead?

File: git.lua
Comment (vladimir-kotikov) [inline git.lua:170]: Nit: it's shorter to define `star` variable as follows
```suggestion
        local star = '*'
        if ver.supports_query_rl_var() and rl.isvariabletrue('colored-stats') then
            star = color.get_clink_color('color.git.star')..'*'..color.get_clink_color('color.filtered')
        end
```

and use it below
```
         :concat(predicted_branches:map(function(branch) return star..branch end))
```

**Review by vladimir-kotikov — CHANGES_REQUESTED** [review]

File: modules/color.lua
Comment (chrisant996) [reply modules/color.lua:29]: It's a complete SGR Parameters sequence, and can include anything the user wants to include, even 24-bit color or underline and so on.

My concern about `colors.color_text("foo", settings.get("color.git.star"))` is that it doesn't encapsulate the version check.  The caller must wrap it with their own version check to avoid blowing up.  If Clink doesn't support color settings, I would like for that to be the same as the user having set the color setting to nothing, and just return the empty string.  Then the caller doesn't have to add a version check -- which is both a convenience and a defensive design: if the caller forgets to check the version, their code will still work correctly.

I'm happy to rename `color.get_clink_color` if another name is preferred, but I think its functional contract is the right one to have, and I don't want to change that part.

File: modules/version.lua
Comment (chrisant996) [reply modules/version.lua:9]: I see pros and cons in both approaches.  E.g. I reasoned that a function is more flexible for the long term, and would avoid a situation where someday some version checks can be variables but some must be functions.  Inconsistency would make it a fragile interface.

I think the execution time optimization there is so small that I would favor instead choosing based on semantic and maintenance considerations, i.e. the ability to prevent future bugs.

But the cons of that approach are not great enough for me to want to spend time on that conversation, so I'll make that change.

File: git.lua
Comment (chrisant996) [reply git.lua:11]: Yes.

File: git.lua
Comment (chrisant996) [reply git.lua:170]: I think reasonable minds can disagree, with good reasons.  I don't want to spend time on that debate.  :)

File: modules/version.lua
Comment (chrisant996) [reply modules/version.lua:9]: I would like to request some room for stylistic and philosophical differences of opinion.  It can become frustrating if stylistic changes keep requiring verification passes.  That's a hidden cost that I would like to keep relatively small.

**Comment by chrisant996** [comment]
> I'd rather just ignore the coverage and just add tests/mock things only if you feel that it brings any real value and doesn't worth too much time

Ok, sounds good.  Now that I've spent the coding time and validation time on the change to trick the code coverage, I would like to keep it as is.  Both approaches function exactly the same, one is just an alternative way to write the expression on a single line.

**Comment by chrisant996** [comment]
I think I got everything updated.

File: modules/color.lua
Comment (vladimir-kotikov) [reply modules/color.lua]: Nah, since `settings.get` would return a complete EGR sequence `color_text` wouldn't work - I just feel like there could be a better/more descriptive/less specific name rather than `get_clink_color` but I can't come up with the better option ATM os let's leave it as is

**Comment by vladimir-kotikov** [comment]
Awesome! Apologize for being so nitpicky and thank you so much for your efforts! - looks like the PR is good to go now )

**Pull request merged**
